### PR TITLE
PKCS#11: Document SoftHSM2 remote usage

### PR DESCRIPTION
**Pull request by cmouse**
### Short description
Fixes a documentation mistake, and also documents how to use SoftHSM2 token remotely. This makes SoftHSM2 integration more secure by keeping the actual token out of PowerDNS process.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ ] compiled this code
- [ ] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by Habbie** [comment]
dnsdist CI failure is unrelated

File: docs/dnssec/pkcs11.rst [hunk 1]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
     pdnsutil show-zone example.com
 
+SoftHSM2 with forwarding
+------------------------
+
+Based on https://p11-glue.github.io/p11-glue/p11-kit/manual/remoting.html.
+
+You need to install ``gnutls-bin`` to get token URLs.
+
+You cannot run ``p11-kit server`` as root, so you will need some user for running it. This user must belong to ``softhsm`` group.
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:50]:
```suggestion
You cannot run ``p11-kit server`` as root, so you will need some user for running it. This user must be in the ``softhsm`` group.
```

File: docs/dnssec/pkcs11.rst [hunk 2]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
+
+These commands need to be ran as the non-root user (we shall call it tokenuser)
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:52]:
```suggestion
These commands need to be run as the non-root user (we shall call it ``tokenuser``).
```

File: docs/dnssec/pkcs11.rst [hunk 3]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
+
+First, setup your token::
+
+   softhsm2-util --init-token --label "ecdsa#1" --pin 1234 --so-pin 1234 --free
+   pkcs11-tool --module /usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so --keypairgen --key-type EC:prime256v1 --pin 1234 -a 'my key'
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:57]: How does this command know it should use the token you inited one line earlier? Am I missing something?

File: docs/dnssec/pkcs11.rst [hunk 4]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
+
+Based on https://p11-glue.github.io/p11-glue/p11-kit/manual/remoting.html.
+
+You need to install ``gnutls-bin`` to get token URLs.
+
+You cannot run ``p11-kit server`` as root, so you will need some user for running it. This user must belong to ``softhsm`` group.
+
+These commands need to be ran as the non-root user (we shall call it tokenuser)
+
+First, setup your token::
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:54]:
```suggestion
First, set up your token::
```

File: docs/dnssec/pkcs11.rst [hunk 5]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
+
+   softhsm2-util --init-token --label "ecdsa#1" --pin 1234 --so-pin 1234 --free
+   pkcs11-tool --module /usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so --keypairgen --key-type EC:prime256v1 --pin 1234 -a 'my key'
+
+Ensure it's there::
+
+   pkcs11-tool --module /usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so -l -O -p 1234
+
+Setup forwarding::
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:63]:
```suggestion
Set up forwarding::
```

File: docs/dnssec/pkcs11.rst [hunk 6]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
+
+  cat <<EOF > /etc/pkcs11/modules/p11-kit-client.module
+  module: /usr/lib/x86_64-linux-gnu/pkcs11/p11-kit-client.so
+  EOF
+
+  p11-kit server -u pdns --provider /usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so "pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=29fdc44dc0d61539;token=ecdsa%231"
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:69]: should we point out that people can't just copy this serial?

File: docs/dnssec/pkcs11.rst [hunk 7]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
+    token model        : SoftHSM v2
+    token flags        : login required, rng, token initialized, PIN initialized, other flags=0x20
+    hardware version   : 2.5
+    firmware version   : 2.5
+    serial num         : 29fdc44dc0d61539
+    pin min/max        : 4/255
+
+Then assign the HSM token to your zone with::
+
+  ~/pdns/bin/pdnsutil hsm assign hsm.test ecdsa256 ksk p11-kit-client 'ecdsa#1' 1234 'my key'
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:89]:
```suggestion
  pdnsutil hsm assign example.com ecdsa256 ksk p11-kit-client 'ecdsa#1' 1234 'my key'
```

File: docs/dnssec/pkcs11.rst [hunk 8]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
+
+And then verify with ``show-zone`` that now the zone has valid key.
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:91]:
```suggestion
And then verify with ``show-zone`` that the zone now has a valid key.
```

File: docs/dnssec/pkcs11.rst [hunk 9]
```diff
@@ -40,6 +40,67 @@ These instructions have been tested on Debian 10 (Buster).
...
+You can do this over SSH as well::
+
+    ssh -R /var/run/pdns/pkcs11:${P11_KIT_SERVER_ADDRESS#*=} pdns@server
+    export P11_KIT_SERVER_ADDRESS=/var/run/pdns/pkcs11
+
+Verify that the token is visible::
+
+   pkcs11-tool --module /usr/lib/x86_64-linux-gnu/pkcs11/p11-kit-client.so -T 
+
+Then use the hsm assign command like before to assign key to your zone, and now you have DNSSEC over SSH.
```
Comment (Habbie) [inline docs/dnssec/pkcs11.rst:102]:
```suggestion
Then use the ``pdnsutil hsm assign`` command like before to assign the key to your zone; now you have DNSSEC over SSH.
```

File: docs/dnssec/pkcs11.rst (hunk 3 above)
Comment (cmouse) [reply docs/dnssec/pkcs11.rst:57]: Well, it works if you only have one slot in use. But I guess I could figure out the syntax for specifying slot too.

File: docs/dnssec/pkcs11.rst (hunk 6 above)
Comment (cmouse) [reply docs/dnssec/pkcs11.rst:69]: You have to copy the entire URL. I should point this out.

**Comment by Habbie** [comment]
Can be merged when the spellcheck is corrected.

**Comment by Habbie** [comment]
> Can be merged when the spellcheck is corrected.

No failure from it. (no report from it, even). Going to merge and see.

**Pull request merged**